gadget/install/install.go

```diff
@@ -306,21 +307,86 @@ func FactoryReset(model gadget.Model, gadgetRoot, kernelRoot, bootDevice string,
 		return nil, fmt.Errorf("cannot read %v partitions: %v", bootDevice, err)
 	}
 
-	// verify that layout is compatible, note this will fail if partition
-	// table is corrupted and the expected partitions are not present
+	roleNeedsEncryption := func(role string) bool {
+		return role == gadget.SystemData || role == gadget.SystemSave
+	}
+
 	layoutCompatOps := &gadget.EnsureLayoutCompatibilityOptions{
 		AssumeCreatablePartitionsCreated: true,
+		ExpectedStructureEncryption:      map[string]gadget.StructureEncryptionParameters{},
+	}
+	if options.EncryptionType != secboot.EncryptionTypeNone {
+		var encryptionParam gadget.StructureEncryptionParameters
+		switch options.EncryptionType {
+		case secboot.EncryptionTypeLUKS:
+			encryptionParam = gadget.StructureEncryptionParameters{Method: gadget.EncryptionLUKS}
+		default:
+			// XXX what about ICE?
+			return nil, fmt.Errorf("unsupported encryption type %v", options.EncryptionType)
+		}
+		for _, volStruct := range laidOutBootVol.LaidOutStructure {
+			if !roleNeedsEncryption(volStruct.Role) {
+				continue
+			}
+			if layoutCompatOps.ExpectedStructureEncryption == nil {
+				layoutCompatOps.ExpectedStructureEncryption = map[string]gadget.StructureEncryptionParameters{}
+			}
+			layoutCompatOps.ExpectedStructureEncryption[volStruct.Name] = encryptionParam
+		}
 	}
+	// factory reset is done on a system that was once installed, so this
+	// should be always successful unless the partition table has changed
 	if err := gadget.EnsureLayoutCompatibility(laidOutBootVol, diskLayout, layoutCompatOps); err != nil {
 		return nil, fmt.Errorf("gadget and system-boot device %v partition table not compatible: %v", bootDevice, err)
 	}
 
-	hasSavePartition := len(partitionsWithRolesAndContent(laidOutBootVol, diskLayout, []string{gadget.SystemSave})) != 0
+	var keyForRole map[string]keys.EncryptionKey
+	var deviceForRole = map[string]string{}
+
+	savePart := partitionsWithRolesAndContent(laidOutBootVol, diskLayout, []string{gadget.SystemSave})
+	hasSavePartition := len(savePart) != 0
+	if hasSavePartition {
+		deviceForRole[gadget.SystemSave] = savePart[0].Node
+	}
 	rolesToReset := []string{gadget.SystemBoot, gadget.SystemData}
 	partsToReset := partitionsWithRolesAndContent(laidOutBootVol, diskLayout, rolesToReset)
 	for _, part := range partsToReset {
```

Comment:
how different is this loop from the one in install? from a quick scan they seem very similar? can we share more code?

Reply:
I've split out all of the bits related to encrypting device, mkfs and populating the partition with content.

Reply:
thanks